crates/cargo-remove/tests/testsuite/cargo_remove/remove_avoid_empty_tables/stderr.log

```diff
@@ -1 +1,2 @@
     Removing clippy from dependencies
+    Updating `dummy-registry` index
```

Comment:
Thats unfortunate

In theory, we don't need to update the index since no additions were made.  I wonder if we can force the Config into offline mode.

Reply:
I will look into that.

Reply:
Okay, I've poked around a bit trying to find a nice way to solve this, and I'm not seeing much. Simply forcing offline mode causes resolution errors, so we'd have to do something more nuanced. Any ideas?

Reply:
Okay, I've poked around a bit trying to find a nice way to solve this, and I'm not seeing much. Simply forcing offline mode causes resolution errors, so we'd have to do something more nuanced. Any ideas?

Reply:
That seems strange.  Cargo should only need the network if there is something unknown in the lock file but a removal shouldn't be unknown.


In testing it outside of cargo, it seems to work
```$
$ cargo metadata --offline  # works
$ ... removed a dependency
$ cargo metadata --offline  # still works
```

Reply:
Ah, I think the problem is that in the test environments, the dependencies *are* unknown. Running `cargo remove` in a normal environment does not result in an index update as expected. Is this something we should try to take into account in the tests?

Reply:
We should be able to generate a valid input lock file for the tests to work successfully.  It is a bit of a pain though (had to do it for cargo-add)

Reply:
Can you point to which test(s) you're referring to? I see a bunch which do update the index.

Reply:
https://github.com/rust-lang/cargo/pull/10902/files#diff-6d9f675d3aeac9ea92f9d02b9f56f709da8c25013f9449c5ef61b913dcaee072

Reply:
Okay! That test still updates the index though... I think I'm a bit confused as to what your point is.

Not sure if you saw it, but the lock file test added in 599854a works just fine, with the help of 5858c88.

Reply:
> Okay! That test still updates the index though... I think I'm a bit confused as to what your point is.

`cargo rm` should not be doing index updates.  

You had commented on the reason

> Ah, I think the problem is that in the test environments, the dependencies are unknown. Running cargo remove in a normal environment does not result in an index update as expected. Is this something we should try to take into account in the tests?

I took that to mean that the `.in/Cargo.lock` is not correct, causing cargo to re-evaluate it and check the index (a problem I ran into when writing cargo add tests).  I then gave the linked cargo add test as an example of one with a valid lock file.  However, the test will need to access the registry because it is adding a dependency and the lock file generator will need to generate new entries.

Reply:
> In theory, we don't need to update the index since no additions were made.

In simple cases, yes. However, there are a few edge cases in which we don't know enough information to correctly modify the `Cargo.lock`, and therefore need access to metadata contained within registry. For example, with features:

```toml
[dependencies]
serde = { version = "1.0.144", features = ["derive"] }
serde_json = "1.0.85"
```

If we `cargo remove serde`, then in the lock file:
- `serde` is removed from the package's dependencies, but not from the file.
- `serde_derive` is removed from the lock file and from `serde`'s dependencies.

A naive solution, a subset of `resolve_ws` which does its best to prune the lock file, would fail to remove `serde_derive`.

We could still implement this naive solution and be careful to call it only when we are certain that we are not in one of these edge cases (or otherwise default to `resolve_ws`). I'm not a fan of this solution, since it means we need to handle each of these edge cases as they come, and as such it's not particularly future-proofed.

We could also *always* opt for this workaround, making conservative changes and letting another cargo subcommand clean up after ourselves. Not a huge fan of this either, since it provides marginal benefit to the user.

Practically speaking, I think it's relatively rare that registry indexes will be updated on an end user's machine. These unit tests are unrepresentative of that environment, since they don't have a cache already set up. As such, this feels more like a non-issue to me. Thoughts?

Reply:
Context: Ed has asked in a separate communication channel for more context on this issue.

Here's what I found working through the stacktrace (in reverse):
1. [`resolve_ws`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/ops/resolve.rs#L65): resolve workspace dependencies using registry inferred by config
2. [`resolve_with_registry`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/ops/resolve.rs#L195): resolve again last lockfile and write result
3. [`resolve_with_previous`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/ops/resolve.rs#L456): resolution internal implementation, adds
   preferences for previous versions
4. [`resolve`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/resolver/mod.rs#L140): perform the resolution, returning the resultant dependency graph
5. [`activate_deps_loop`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/resolver/mod.rs#L204): recursively "activate" all dependencies
6. [`activate`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/resolver/mod.rs#L666): pull information ("summary") about dependency from registry
7. [`RegistryQueryer::build_deps`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/resolver/dep_cache.rs#L238): figure out which dependencies are required by a candidate
8. [`RegistryQueryer::query`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/resolver/dep_cache.rs#L105): query the registry for the dependency, first checking cache
9. [`PackageRegistry::query`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/registry.rs#L633): find dependency
10. [`PackageRegistry::ensure_loaded`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/registry.rs#L187): self-explanatory
11. [`PackageRegistry::block_until_ready`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/registry.rs#L730): try to download the dependency
12. [`ReplacedSource::block_until_ready`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/sources/replaced.rs#L127): ...
13. [`RegistrySource::block_until_ready`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/sources/registry/mod.rs#L825): ...
14. [`RemoteRegistry::block_until_ready`](https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/sources/registry/remote.rs#L255): actually download the dependency

This process exists in order to have sufficient information about a dependency (i.e.a summary) to resolve it. In finding this, at step 8 the cache is checked.

This is what is contained in the summary and required for dependency resolution:

https://github.com/rust-lang/cargo/blob/646e9a0b9ea8354cc409d05f10e8dc752c5de78e/src/cargo/core/summary.rs#L12-L28

Reply:
Offline test added in 078968b.